### ISSUE TITLE
Screenshots sometimes out-of-sync with actual state of web contents

### ISSUE THREAD

**Issue by Mr0grog** [segmentio/nightmare#468]
In trying to use Nightmare for doing some screen captures of highly dynamic web pages, I've discovered that the screenshots are often out-of-sync with the state of the web page being captured. After a lot of investigation, it turns out that this is actually an underlying issue in Electron—it generates a screenshot by simply copying from Chrome's backing store, which is not actually guaranteed to be up-to-date (see [Electron issue 1049](https://github.com/atom/electron/issues/1049) and [this Chromium-dev discussion](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-dev/o2aUtFAIlyQ)).

I believe this is probably the underlying issue behind a few other nightmare bugs, such as: https://github.com/segmentio/nightmare/issues/33, https://github.com/segmentio/nightmare/issues/129, the long wait @rosshinkley suggests adding in his solution to https://github.com/segmentio/nightmare/issues/328

But! It turns out there is a way to address this. In Electron's `BrowserWindow.webContents`, you can call `beginFrameSubscription(callback)`, which seems (in my testing) to guarantee that the backing store is up-to-date when your callback is called. So if you changed [line 252 of `lib/runner.js`](https://github.com/segmentio/nightmare/blob/master/lib/runner.js#L252) to:

```js
win.webContents.beginFrameSubscription(function(buffer) {
  win.webContents.endFrameSubscription();
  win.capturePage.apply(win, args);
});
```

You ought to have a much more deterministic result when calling `.screenshot()`.

Unfortunately, I don't have a great reduced test case for this, but I'm more than happy to submit a patch if desired.

**Comment by matthewmueller** [comment]
awesome! we'd love to see a patch for a more deterministic result!

**Comment by rosshinkley** [comment]
@Mr0grog Glad to finally have an answer to that question, and good eye for finding that issue!  Especially good to know it's not a framebuffer-dependent issue.

I agree: anything to make the behavior deterministic.  I do have an edge-case question, though - in my read of the documentation, the `beginFrameSubscription` callback is called for any display event.  If the screenshot is taken well after the page is rendered, will the callback still be called?

**Comment by Mr0grog** [comment]
@matthewmueller Great, I'll try and put together a PR this week.

@rosshinkley That is a really good question! I will poke around and see what I can see. In the worst case, twiddling with something on the page, such as adding and removing an element, ought to work to trigger it.

**Comment by Mr0grog** [comment]
Spent a little time on this today and discovered:

1. If the page is all settled and idle, the callback won't be fired (as you surmised @rosshinkley), so we do have to do something to the page to trigger it. That something has to affect the render tree—adding comment nodes, empty text nodes, etc. won't do it. Happily, adding a transparent 1x1 div to the top left corner of the page (even if the top/left corner isn't in view!) works.

  (Also interesting: even though adding it to the top of the page when the top of the page isn't visible does work adding it where it never *could* be drawn, e.g. `left: -10000px` does *not* work. ¯\_(ツ)_/¯ )

2. It's easy to crash electron while doing this (see https://github.com/atom/electron/issues/4441), so the whole technique may be a bit problematic in the end. If we stay subscribed to frames for the lifetime of the window, I think we could reliably avoid the crash, but I can imagine that easily getting resource intensive.

**Comment by Zn4rK** [comment]
@Mr0grog Would [window.requestAnimationFrame()](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) trigger the render tree?

**Comment by Mr0grog** [comment]
It certainly triggers the render tree, but this isn't actually a render tree issue. Also, I tried the same thing when I was first exploring this ;)

It's fundamentally an issue with how Chrome's *backing store* gets updated, which is actually largely unrelated to what is getting rendered to the screen inside the renderer process (where the web page is being drawn to the screen). In any case, there is a complete fix for this issue in #479, it just needs reviewing.

**Comment by rosshinkley** [comment]
#479 has been merged, closing.
